sources/io/print.dylan

```diff
@@ -1326,18 +1326,41 @@ define method do-printing-object
         write(stream, as-lowercase(as(<byte-string>, class.debug-name)));
         write(stream, " ");
         continuation(stream);
-        //--- write(stream, " ");
-        //--- write(stream, integer-to-string(object-address(object), base: 16));
+        write(stream, " #x");
+        write(stream, integer-to-string(object-id(object), base: 16));
       type? =>
         write(stream, as-lowercase(as(<byte-string>, class.debug-name)));
         write(stream, " ");
         continuation(stream);
       identity? =>
         continuation(stream);
-        //--- write(stream, " ");
-        //--- write(stream, integer-to-string(object-address(object), base: 16));
+        write(stream, " #x");
+        write(stream, integer-to-string(object-id(object), base: 16));
       otherwise =>
         continuation(stream);
     end
   end
 end method do-printing-object;
+
+// Cache of integer IDs we use instead of the object address when printing, in case we're
+// using a moving GC. <object> => <integer>
+define constant $object-id-cache :: <table> = make(<table>, weak: #"key");
+
+define variable *next-object-id* :: <integer> = 1;
+define constant $next-object-id-lock :: <lock> = make(<lock>);
```

Comment:
Instead of a lock, could use `atomic-increment!`

Reply:
Yep, I considered that but couldn't see a way to make it work.